ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilterSource.java

```diff
@@ -0,0 +1,92 @@
+ * Copyright ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods;
+
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.processor.TransactionTrace;
+import org.hyperledger.besu.ethereum.api.util.ArrayNodeWrapper;
+import org.hyperledger.besu.ethereum.core.Block;
+import org.hyperledger.besu.ethereum.core.Transaction;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+
+public class TraceFilterSource implements Iterator<TransactionTrace> {
+
+  private final ArrayNodeWrapper resultArrayNode;
+  private final Iterator<Block> blockIterator;
+  private Iterator<TransactionTrace> transactionTraceIterator;
+  private Block currentBlock;
+
+  public TraceFilterSource(final List<Block> blockList, final ArrayNodeWrapper resultArrayNode) {
+    this.resultArrayNode = resultArrayNode;
+    this.blockIterator = blockList.iterator();
+    this.transactionTraceIterator = getNextTransactionIterator();
+  }
+
+  private Iterator<TransactionTrace> getNextTransactionIterator() {
+    if (blockIterator.hasNext()) {
+      currentBlock = blockIterator.next();
+      if (currentBlock.getBody().getTransactions().size() == 0) {
+        TransactionTrace rewardTrace = new TransactionTrace(Optional.of(currentBlock));
+        return Collections.singleton(rewardTrace).iterator();
+      } else {
+        List<TransactionTrace> transactionTraces = new ArrayList<>();
+        List<Transaction> transactions = currentBlock.getBody().getTransactions();
+        int size = transactions.size();
+        int index = 0;
+        while (index <= size) {
+          if (index == size) {
+            transactionTraces.add(new TransactionTrace(Optional.of(currentBlock)));
+          } else {
+            transactionTraces.add(
+                new TransactionTrace(transactions.get(index), Optional.of(currentBlock)));
+          }
+          index++;
+        }
+        return transactionTraces.iterator();
+      }
+    }
+    return null;
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (resultArrayNode.isFull()) return false;
+    if (transactionTraceIterator == null) {
+      return false;
+    }
+    if (transactionTraceIterator.hasNext()) {
+      return true;
+    }
+    transactionTraceIterator = getNextTransactionIterator();
+    return hasNext();
```

Comment:
if we start with an empty blockIterator, this can infinitely recurse no?  we should assert that the blockList and therefore blockIterator is not empty in the constructor, or change the flow to make sure we cannot get into trouble.

Reply:
Nevermind - it is getting late and I read the first condition as != null.

Reply:
There is no infinite recurse as we have this condition at the beginning   
```
if (!blockIterator.hasNext()) {
      return null;
    } 
```
There're actually unit tests that test this specific use case, a bloc with no transactions. In the unit tests, block 0 doesn't exist and block 1 is a block with 0 transactions.